Caching Terraform providers (#3001)
gruntwork-io/terragrunt: docs/_docs/04_reference/cli-options.md

**Environment Variable**: `TERRAGRUNT_TF_JSON_LOG` (set to `true`)

When this flag is set, Terragrunt will wrap Terraform `stdout` and `stderr` in JSON log messages. Works only with `--terragrunt-json-log` flag.

> ### terragrunt-provider-cache

Reviewer: For each of these new provider caching flags, add a link to the dedicated provider caching docs page!

Author: Added. Thanks.

**CLI Arg**: `--terragrunt-provider-cache`
**Environment Variable**: `TERRAGRUNT_PROVIDER_CACHE`
**Commands**:
- [run-all](#run-all)

> Enables provider caching. This forces Terraform to make provider requests through the built-in Provider Cache server.

Reviewer: ```suggestion
Enables Terragrunt's provider caching. This forces Terraform to make provider requests through the built-in Terragrunt Provider Cache server.
```

Author: Thanks.

### terragrunt-provider-cache-dir

**CLI Arg**: `--terragrunt-provider-cache-dir`
**Environment Variable**: `TERRAGRUNT_PROVIDER_CACHE_DIR`
**Commands**:
- [run-all](#run-all)

> The path to the cache directory. By default, `terragrunt/providers` folder in the user cache directory: `$HOME/.cache` on Unix systems, `$HOME/Library/Caches` on Darwin, `%LocalAppData%` on Windows. The file structure of the cache directory is identical to the Terraform [plugin_cache_dir](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-plugin-cache) directory.

Reviewer: ```suggestion
The path to the Terragrunt provider cache directory. By default, `terragrunt/providers` folder in the user cache directory: `$HOME/.cache` on Unix systems, `$HOME/Library/Caches` on Darwin, `%LocalAppData%` on Windows. The file structure of the cache directory is identical to the Terraform [plugin_cache_dir](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-plugin-cache) directory.
```

Author: Thanks.

### terragrunt-provider-cache-complete-lock

**CLI Arg**: `--terragrunt-provider-cache-complete-lock`
**Environment Variable**: `TERRAGRUNT_PROVIDER_CACHE_COMPLETE_LOCK`
**Commands**:
- [run-all](#run-all)

> Disables Terraform [_plugin_cache_may_break_dependency_lock_file_](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-installation) feature. When this feature is enabled, Terraform generates `.terraform.lock.hcl` files based on the files in the cache directory. If the flag is set Terragrunt disable this feature and before each `init` command runs `terraform providers lock` to generate complete `.terraform.lock.hcl`, which in turn downloads providers from the original registry, which is not fast.

Reviewer: ```suggestion
By default, Terraform does _not_ use the cache for modules without a lock file. This results in lots of extra provider downloading. To work around this, for modules without a lock file, Terragurnt provider caching enables the [_plugin_cache_may_break_dependency_lock_file_](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-installation) feature, which allows Terraform to generate a partial lock file if it finds the providers it needs in the cache. This avoids lots of unnecessary provider downloads, but results in partial lock files. If you wish to disable this feature, set this flag flag, and Terragrunt will run `terraform providers lock` before `init` for modules without lock files, which will generate a complete lock file, but at the cost of more provider downloads.
```

### terragrunt-provider-cache-hostname

**CLI Arg**: `--terragrunt-provider-cache-hostname`
**Environment Variable**: `TERRAGRUNT_PROVIDER_CACHE_HOSTNAME`
**Commands**:
- [run-all](#run-all)

> The hostname of the built-in Terragrunt Cache server. By default, 'localhost'.

Reviewer: ```suggestion
The hostname of the built-in Terragrunt Provider Cache server. By default, 'localhost'.
```

Author: Thanks.

### terragrunt-provider-cache-port

**CLI Arg**: `--terragrunt-provider-cache-port`
**Environment Variable**: `TERRAGRUNT_PROVIDER_CACHE_PORT`
**Commands**:
- [run-all](#run-all)

> The port of the built-in Terragrunt Cache server. By default, assigned automatically.

Reviewer: ```suggestion
The port of the built-in Terragrunt Provider Cache server. By default, assigned automatically.
```

Author: Thanks.

### terragrunt-provider-cache-token

**CLI Arg**: `--terragrunt-provider-cache-token`
**Environment Variable**: `TERRAGRUNT_PROVIDER_CACHE_TOKEN`
**Commands**:
- [run-all](#run-all)

> The Token for authentication on the built-in Terragrunt Cache server. By default, assigned automatically.

Reviewer: ```suggestion
The Token for authentication on the built-in Terragrunt Provider Cache server. By default, assigned automatically.
```

Author: Thanks.

### terragrunt-provider-cache-registry-names

**CLI Arg**: `--terragrunt-provider-cache-registry-names`
**Environment Variable**: `TERRAGRUNT_PROVIDER_CACHE_REGISTRY_NAMES`
**Commands**:
- [run-all](#run-all)

> The list of remote registries to cached by Terragrunt Cache server. By default, 'registry.terraform.io', 'registry.opentofu.org'.

Reviewer: ```suggestion
The list of remote registries to cached by Terragrunt Provider Cache server. By default, 'registry.terraform.io', 'registry.opentofu.org'.
```
